pir/cpp/database.cpp

```diff
@@ -43,19 +44,122 @@ StatusOr<std::shared_ptr<PIRDatabase>> PIRDatabase::Create(
   return std::make_shared<PIRDatabase>(db, std::move(context));
 }
 
-StatusOr<std::vector<seal::Ciphertext>> PIRDatabase::multiply(
-    const std::vector<seal::Ciphertext>& in) {
-  std::vector<seal::Ciphertext> result(in.size());
+using seal::Ciphertext;
+using seal::Evaluator;
+using seal::Plaintext;
+using std::vector;
 
-  for (size_t idx = 0; idx < in.size(); ++idx) {
-    seal::Ciphertext ct;
-    try {
-      context_->Evaluator()->multiply_plain(in[idx], db_[idx], ct);
-    } catch (std::exception& e) {
-      return InvalidArgumentError(e.what());
+// Recursive function to do the dot product of each dimension.
+// Calls itself to move down dimensions until you get to the bottom dimension.
+// Bottom dimension just does a dot product with the DB, and returns the result.
+// Upper levels then take those results, and dot product again with the
+// selection vector, until you get back to the top.
+// NB: the database iterator is passed by reference so we keep track of where
+// we are in the database. Be very careful with that iterator!
+Ciphertext multiply_dims(
+    Evaluator& evaluator, vector<uint32_t> dimensions,
```

Comment:
Wouldn't be easier to read if you pass start indexes and references to the db & selection_vector? instead of the iterators.

Reply:
Not really. The iterator keeps track of where we are within the database. Otherwise, we'd have to keep track of a separate database reference and indices, which gets messier. Here it's very clear where we are in the database. It also doesn't fix the problem of having to pass back how far we've gone in the database.